src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/WPF/ApplicationXamlFileAccessor.cs

```diff
@@ -0,0 +1,190 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using System.Runtime.InteropServices;
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Design.Serialization;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.TextManager.Interop;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.WPF;
+
+[Export(typeof(IApplicationXamlFileAccessor))]
+[AppliesTo(ProjectCapability.DotNet)]
+internal class ApplicationXamlFileAccessor : IApplicationXamlFileAccessor
+{
+    private readonly UnconfiguredProject _project;
+    private readonly IServiceProvider _serviceProvider;
+    private readonly IVsUIService<IVsRunningDocumentTable> _runningDocumentTable;
+    private readonly IProjectThreadingService _threadingService;
+
+    private DocData? _applicationXamlDocdata;
+    private AppDotXamlDocument? _applicationDotXamlDocument;
+
+    [ImportingConstructor]
+    public ApplicationXamlFileAccessor(
+        UnconfiguredProject project,
+        [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
+        IVsUIService<IVsRunningDocumentTable> runningDocumentTable,
+        IProjectThreadingService threadingService)
+    {
+        _project = project;
+        _serviceProvider = serviceProvider;
+        _runningDocumentTable = runningDocumentTable;
+        _threadingService = threadingService;
+    }
+
+    public async Task<string?> GetStartupUriAsync()
+    {
+        await _threadingService.SwitchToUIThread();
```

Comment:
We serialize access to the `_applicationDotXamlDocument` and related data structures by always switching to the UI thread. The `DocData` needs to be accessed from that thread, anyway.